hedera-node/hedera-app/src/main/java/com/hedera/node/app/tss/TssCryptographyManager.java

```diff
@@ -0,0 +1,161 @@
+     * The most expensive operations involving {@link TssLibrary} are executed asynchronously.
+     *
+     * @param tssMessageBodies        the list of TSS messages
+     * @param tssParticipantDirectory the TSS participant directory
+     * @return a CompletableFuture containing the ledger id and signature if the threshold is met, null otherwise
+     */
+    private CompletableFuture<LedgerIdWithSignature> computeAndSignLedgerIdIfApplicable(
+            @NonNull final List<TssMessageTransactionBody> tssMessageBodies,
+            final TssParticipantDirectory tssParticipantDirectory) {
+        return CompletableFuture.supplyAsync(() -> {
+            // Validate TSS transactions and set the vote bit set.
+            final var validTssOps = validateTssMessages(tssMessageBodies, tssParticipantDirectory, tssLibrary);
+            boolean tssMessageThresholdMet = isThresholdMet(validTssOps, tssParticipantDirectory);
+
+            // If the threshold is not met, return
+            if (!tssMessageThresholdMet) {
+                return null;
+            }
+            final var validTssMessages = getTssMessages(validTssOps);
+            final var computedPublicShares = tssLibrary.computePublicShares(tssParticipantDirectory, validTssMessages);
+
+            // compute the ledger id and sign it
+            final var ledgerId = tssLibrary.aggregatePublicShares(computedPublicShares);
+            final var signature = gossip.sign(ledgerId.publicKey().toBytes());
+
+            final BitSet tssVoteBitSet = computeTssVoteBitSet(validTssOps);
+            return new LedgerIdWithSignature(ledgerId, signature, tssVoteBitSet);
+        });
+    }
+
+    /**
+     * Compute the TSS vote bit set. No need to validate the TSS messages here as they have already been validated.
+     *
+     * @param validIssBodies the valid TSS messages
+     * @return the TSS vote bit set
+     */
+    private BitSet computeTssVoteBitSet(@NonNull final List<TssMessageTransactionBody> validIssBodies) {
+        final var tssVoteBitSet = new BitSet();
+        for (TssMessageTransactionBody op : validIssBodies) {
+            tssVoteBitSet.set((int) op.shareIndex());
+        }
+        return tssVoteBitSet;
+    }
+
+    /**
+     * Check if the threshold is met. The threshold is met if more than half the consensus weight has been received.
+     *
+     * @param validTssMessages        the valid TSS messages
+     * @param tssParticipantDirectory the TSS participant directory
+     * @return true if the threshold is met, false otherwise
+     */
+    private boolean isThresholdMet(
+            @NonNull final List<TssMessageTransactionBody> validTssMessages,
+            @NonNull final TssParticipantDirectory tssParticipantDirectory) {
+        final var numShares = tssParticipantDirectory.getShareIds().size();
+        // If more than 1/2 the consensus weight has been received, then the threshold is met
+        return validTssMessages.size() >= ((numShares + 2) / 2);
```

Comment:
This threshold value is calculated in a couple of places; may be worth extracting it to a dedicated method.

Reply:
Agreed. Fixed in both places.